app/uk/gov/hmrc/customs/declarations/information/services/declarationsInformationServices.scala

```diff
@@ -196,51 +197,58 @@ abstract class DeclarationService @Inject()(override val apiSubFieldsConnector:
-            Left(customNotFoundResponse.XmlResult.withConversationId)
-          case e: HttpException =>
-            logger.warn(s"declaration [$endpointName] call failed with backend http status code of [${e.responseCode}]: [${e.getMessage}] so returning 500 to consumer")
-            Left(ErrorResponse.ErrorInternalServerError.XmlResult.withConversationId)
-          case _: CircuitBreakerOpenException =>
-            logger.error("unhealthy state entered so returning 500 to consumer with message service unavailble")
-            Left(errorResponseServiceUnavailable.XmlResult.withConversationId)
-          case NonFatal(e) =>
-            logger.error(s"declaration [$endpointName] call failed: [${e.getMessage}] so returning 500 to consumer", e)
-            Left(ErrorResponse.ErrorInternalServerError.XmlResult.withConversationId)
-        }
+          })
+          .recover(recoverException(asr, hc))
       case Left(result) =>
         Future.successful(Left(result))
     }
   }
 
-  def processError(errorResponse: ErrorResponse)(implicit r: HasConversationId): Left[Result, Nothing] = {
-    logFailureOutcome(errorResponse)
+  private def recoverException[A](implicit asr: AuthorisedRequest[A], hc: HeaderCarrier): PartialFunction[Throwable, Either[Result, HttpResponse]] = {
+    case e: Non2xxResponseException if e.responseCode == INTERNAL_SERVER_ERROR =>
+      returnErrorResult(asr, (XML.loadString(e.response.body) \ "errorCode").text)
+
+    case e: Non2xxResponseException if e.responseCode == FORBIDDEN && config.informationConfig.payloadForbiddenEnabled =>
+      returnErrorResult(asr, e, FORBIDDEN, ErrorPayloadForbidden)
+
+    case e: Non2xxResponseException if e.responseCode == FORBIDDEN =>
+      returnErrorResult(asr, e, INTERNAL_SERVER_ERROR, ErrorInternalServerError)
+
+    case e: HttpException if e.responseCode == NOT_FOUND =>
+      returnErrorResult(asr, e, INTERNAL_SERVER_ERROR, customNotFoundResponse)
+
+    case e: HttpException =>
+      returnErrorResult(asr, e, INTERNAL_SERVER_ERROR, ErrorInternalServerError)
+
+    case _: CircuitBreakerOpenException =>
+      returnErrorResult(asr, errorResponseServiceUnavailable, "unhealthy state entered so returning 500 to consumer with message service unavailable", null)
+
+    case NonFatal(e) =>
+      returnErrorResult(asr, ErrorInternalServerError, s"declaration [$endpointName] call failed: [${e.getMessage}] so returning 500 to consumer", e)
+  }
+
+  private def returnErrorResult[A](implicit asr: AuthorisedRequest[A], e: HttpException, returnCode: Int, errorResponse: ErrorResponse): Left[Result, Nothing] = {
+    logger.warn(s"declaration [$endpointName] call failed with backend http status code of [${e.responseCode}]: [${e.getMessage}] so returning [$returnCode] to consumer")
+    Left(errorResponse.XmlResult.withConversationId)
+  }
+
+  private def returnErrorResult[A](implicit asr: AuthorisedRequest[A], errorCodeText: String): Left[Result, Nothing] = {
```

Comment:
Perhaps with an Option 3rd parameter of a name like 'includeResponseBody'

Reply:
@drewmiley it's not just only the response body is the extra also this part of message is different than the another message. "500 and error: [${errorResponse.errorCode}] so returning to consumer response status". It will make the code a bit harder to read if combine plus there will be everywhere with that 'includeReponseBody' What do you think?

Reply:
@drewmiley I've combined them, I found hard to figure out what's the output for the error message. That was the reason why I didn't combine them. Unless you have better idea.
```  
private def recoverException[A](implicit asr: AuthorisedRequest[A], hc: HeaderCarrier): PartialFunction[Throwable, Either[Result, HttpResponse]] = {
    case e: Non2xxResponseException if e.responseCode == INTERNAL_SERVER_ERROR =>
      returnErrorResult(asr, e, INTERNAL_SERVER_ERROR, matchErrorCode((XML.loadString(e.response.body) \ "errorCode").text), true)

    case e: Non2xxResponseException if e.responseCode == FORBIDDEN && config.informationConfig.payloadForbiddenEnabled =>
      returnErrorResult(asr, e, FORBIDDEN, ErrorPayloadForbidden, false)

    case e: Non2xxResponseException if e.responseCode == FORBIDDEN =>
      returnErrorResult(asr, e, INTERNAL_SERVER_ERROR, ErrorInternalServerError, false)

    case e: HttpException if e.responseCode == NOT_FOUND =>
      returnErrorResult(asr, e, INTERNAL_SERVER_ERROR, customNotFoundResponse, false)

    case e: HttpException =>
      returnErrorResult(asr, e, INTERNAL_SERVER_ERROR, ErrorInternalServerError, false)

    case _: CircuitBreakerOpenException =>
      logger.error("unhealthy state entered so returning 500 to consumer with message service unavailable")
      Left(errorResponseServiceUnavailable.XmlResult.withConversationId)

    case NonFatal(e) =>
      logger.error(s"declaration [$endpointName] call failed: [${e.getMessage}] so returning 500 to consumer", e)
      Left(ErrorResponse.ErrorInternalServerError.XmlResult.withConversationId)
  }

  private def returnErrorResult[A](implicit asr: AuthorisedRequest[A], e: HttpException, returnCode: Int, errorResponse: ErrorResponse, includeResponseBody: Boolean): Left[Result, Nothing] = {
    val statusCode: String = if (includeResponseBody) s"500 and error: [${errorResponse.errorCode}]" else s"[${e.responseCode}]: [${e.getMessage}]"
    val returnCodeStr: String = if (includeResponseBody) "" else s"[$returnCode]"
    val responseBody: String = if (includeResponseBody) s"response status [${errorResponse.httpStatusCode}] and response body: [${errorResponse.XmlResult.body.asInstanceOf[HttpEntity.Strict].data.utf8String}]" else ""
    logger.warn(s"declaration [$endpointName] call failed with backend http status code of $statusCode so returning $returnCodeStr to consumer $responseBody")
    Left(errorResponse.XmlResult.withConversationId)
  }
```

Reply:
>  it's not just only the response body is the extra also this part of message is different than the another message. "500 and error: [${errorResponse.errorCode}] so returning to consumer response status".
I think that the fact they differ here is probably an oversight on the part of past development.

Good work though - I'll have a play around with it tomorrow and see if I can make that third param a bit tidier :)